### DISCUSSION TITLE
Use only 2 lines instead 4

### DISCUSSION THREAD

**Discussion by sommo** [micbay/RaceTimerAndController#10] [Q&A] [answered]
Hi @micbay,
first of all, amazing project!
Is it possible to use only 2 line? can you point me in which parts of the code i need to declare that i will use only 2 line?

Thank you

**Comment by micbay** [comment]
Hi sommo,
I'm not sure if you asking if it can be used with only 2 lanes/racers, or used with only 1 racer, using only lane 2. The answer is that both are possible. Which lanes/lines/racers that are active for any given race can be changed and set from the '[Settings Menu](https://github.com/micbay/RaceTimerAndController/blob/master/Images/ScreenShots/Settings_Menu.png)' screen. From this menu screen pressing a number key 1-4 will toggle the active status of that lane #. The 'skull' icon indicates the lanes is disabled. There are no changes to code required to operate in this manner. If you only want to use only lane 2 then you can simply disable lanes, 1, 3, and 4.

If you want different lanes, than 1 & 2, active on bootup, then you can create a 'local.settings.h' file, then edit the following setting.
original default, lanes 1-2 active on bootup: `#define DEFAULT_LANES_ENABLED { Off, StandBy, StandBy, Off, Off }`
example with only lane 2 active on bootup: `#define DEFAULT_LANES_ENABLED { Off, Off, StandBy, Off, Off }`
Note that the first 'Off' term in the array is not used, lanes 1-4 are set by array elements 1-4.
See the end of the readme documentation for more explanation on creating and editing default settings using a 'local.settings.h' file.

Let me know if this isn't what you were talking about,
hope you enjoy the controller,

**Comment by sommo** [comment]
Hi micbay,Thank you for your fast reply, I have a carrera slotrace with only 2 lines so my question is how to compile the sketch with only 2 line active cause I don’t have the 3rd and 4th line (physically). I’m thinking that doing so because maybe and probabily the arduino will be more “light” (talking about load and cpu/memory use because it has to check only 2 line for time/lap. It’s a suggestion anyway.I have also some other question (about adding a stoplight, with red and green led and if your sketch storage fastest lap/record also after a power off or maybe, that’s the point, if its possible to add a sd module to storage this information to get them on future, but le me know if it’s better to open a new thread and treat those argument separately).Thank you againIl giorno 2 mar 2023, alle ore 14:45, micbay ***@***.***> ha scritto:﻿
Hi sommo,
I'm not sure if you asking if it can be used with only 2 lanes/racers, or used with only 1 racer, using only lane 2. The answer is that both are possible. Which lanes/lines/racers that are active for any given race can be changed and set from the 'Settings Menu' screen. From this menu screen pressing a number key 1-4 will toggle the active status of that lane #. The 'skull' icon indicates the lanes is disabled. There are no changes to code required to operate in this manner. If you only want to use only lane 2 then you can simply disable lanes, 1, 3, and 4.
If you want different lanes, than 1 & 2, active on bootup, then you can create a 'local.settings.h' file, then edit the following setting.
original default, lanes 1-2 active on bootup: #define DEFAULT_LANES_ENABLED { Off, StandBy, StandBy, Off, Off }
example with only lane 2 active on bootup: #define DEFAULT_LANES_ENABLED { Off, Off, StandBy, Off, Off }
Note that the first 'Off' term in the array is not used, lanes 1-4 are set by array elements 1-4.
See the end of the readme documentation for more explanation on creating and editing default settings using a 'local.settings.h' file.
Let me know if this isn't what you were talking about,
hope you enjoy the controller,

**Reply by micbay** [reply]
Regarding the lane count, the variable '`laneCount`' sets how many lanes are possible in code. You can try just changing that value, however, the code has not be refactored to accommodate changing the lane count. Many of the data arrays are initialized assuming there are 4 possible lanes so this will almost assuredly cause a compiler error. To fix array size mismatches, one would need to go through everywhere 'laneCount' is used and make sure that any data arrays defined using 'laneCount' will have the proper number of elements, by deleting elements at indexes array[3] and array[4] (ie elements 4 and 5 of the arrays, remember the zero index exists, but is skipped for racer data).

Most likely the only noticeable difference of changing laneCount to 2, would be an ability to increase the number of lap times stored. Loop execution is so much faster than physical changes in the system that I don't think it will have any impact on lap detection and race quality.

The long term plan was to initialize array's dynamically such that they are natively built to match the 'laneCount' variable, but this creates some challenges and increases the code base size to implement extra code to do this. So it's unknown if there would actually be any memory saved or if it would just be trading tit for tat. Yes, loop execution would iterate through fewer lanes, but in most cases only the active lanes are processed in heavily used loops during a race, so I don't think you would find any practical advantage in that regard.

Regarding feature suggestions, it's probably best to create a new discussion for each idea so they can be individually addressed, but to give some immediate feedback on your two ideas:

A Red-Yellow-Green stop light - There are no pins left on the nano to add more LEDs, but you might be able to use an I2C compatible stop light and add it to the 7-segment display bar chain, then add a line of code to the pre-start countdown to operate it.

Non-Volatile memory - It is possible to use the built in EEPROM, used by the Arduino to store the program itself, to also store data to be available after a power cycle, but an EPROM does not have unlimited write cycles, it will eventually fail. There may also be total space limits that cannot be overcome after moving dynamic memory to the EEPROM. An external SD is also a possibility, but I have not done that with Arduino and don't know if it's feasible or not in this project. Read/write times might be an issue if they are relatively slow compared to loop execution time.

Something to keep in mind is that, this project is demanding a lot out of our little processor and is near its max operational memory, and program memory limits. As such, changes that require extra code, that can't be balanced with refactoring existing code, may not be practically possible without some trade-offs in other existing functionality. That's not to say new ideas can't be implemented, but just to say that this project is at a stage that new features may require a bit more finesse and effort to make fit, and at some point it may be best to create a use case specific fork rather than adding to the general code base.

**Comment by sommo** [comment]
Hi @micbay,
thank you again for your complete and exhaustive explanation.
i will open newone discussion for about the Red-Yellow-Green stop light.
Best regards

**Comment by micbay** [comment, marked as answer]
With [version 1.1.0](https://github.com/micbay/RaceTimerAndController/releases/tag/v1.1.0) (which is equivalent to pulling the current master branch at the time of this update), the code base now supports setting the number of lanes that will be used in the system.

The `laneCount` variable is now set using a new `LANE_COUNT` macro id in the `defaultSettings.h`. This is also now set to 2 by default, but is changeable to any value 1-4 by setting an alternate `LANE_COUNT` in ones `localSettings.h`. Subsequently, the new `DEFAULT_MAX_STORED_LAPS` is now 20 out of the box. If `LANE_COUNT` is increased, `DEFAULT_MAX_STORED_LAPS` should be reduced accordingly.

**Reply by sommo** [reply]
wow, i will try asap, thank you!
